reasoner/Reasoner.java

```diff
@@ -77,56 +83,102 @@ ResolverRegistry resolverRegistry() {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
     }
 
-    private Producer<ConceptMap> producer(Conjunction conjunction, Set<Identifier.Variable.Name> filter,
-                                          Context.Query context) {
-        Producer<ConceptMap> producer;
-        logicMgr.typeResolver().resolve(conjunction);
-        if (conjunction.isSatisfiable()) {
-            if (isInfer(context)) producer = resolve(conjunction, filter);
-            else producer = traversalEng.producer(
-                    conjunction.traversal(filter), context.producer(), PARALLELISATION_FACTOR
-            ).map(conceptMgr::conceptMap);
-        } else if (!conjunction.isBounded() && conjunctionContainsThings(conjunction, filter)) {
-            throw GraknException.of(UNSATISFIABLE_CONJUNCTION, conjunction);
+    private ResourceIterator<ConceptMap> resolve(Disjunction disjunction, Set<Identifier.Variable.Name> filter, Long offset, Long limit, Context.Query context) {
+        Set<Reference.Name> nameFilter = iterate(filter).map(Identifier.Variable.Name::reference).toSet();
+        if (disjunction.conjunctions().size() == 1) {
+            return produce(new ReasonerProducer(disjunction.conjunctions().get(0), resolverRegistry, nameFilter, offset, limit), context.producer(), asyncPool1());
         } else {
-            return Producers.empty();
+            return produce(new ReasonerProducer(disjunction, resolverRegistry, nameFilter, offset, limit), context.producer(), asyncPool1());
+        }
+    }
+
+    private Disjunction resolveTypesAndFilter(Disjunction disjunction, Set<Identifier.Variable.Name> filter) {
+        List<Conjunction> satisfiableConjunctions = new ArrayList<>();
+        for (Conjunction conjunction : disjunction.conjunctions()) {
+            logicMgr.typeResolver().resolve(conjunction);
+            if (!conjunction.isSatisfiable() && !conjunction.isBounded() && conjunctionContainsThings(conjunction, filter)) {
+                throw GraknException.of(UNSATISFIABLE_CONJUNCTION, conjunction);
+            }
+            if (conjunction.isSatisfiable()) {
+                Set<Negation> filteredNegations = new HashSet<>();
+                for (Negation negation : conjunction.negations()) {
+                    Set<Identifier.Variable.Name> f = set(filter, iterate(conjunction.variables())
+                            .filter(v -> v.id().isName()).map(v -> v.id().asName()).toSet());
+                    Disjunction resolvedAndFiltered = resolveTypesAndFilter(negation.disjunction(), f);
+                    if (!disjunction.conjunctions().isEmpty()) filteredNegations.add(new Negation(resolvedAndFiltered));
+                }
+                satisfiableConjunctions.add(new Conjunction(conjunction.variables(), filteredNegations));
+            }
         }
+        return new Disjunction(satisfiableConjunctions);
+    }
```

Comment:
recursively condense a query to only the satisfiable components, applying type resolver and validating them as we recurse.